sklearn/tree/_splitter.pyx

```diff
@@ -157,62 +205,18 @@ cdef class Splitter:
         self.y = y
 
         self.sample_weight = sample_weight
-        return 0
-
-    cdef int node_reset(self, SIZE_t start, SIZE_t end,
-                        double* weighted_n_node_samples) nogil except -1:
-        """Reset splitter on node samples[start:end].
-
-        Returns -1 in case of failure to allocate memory (and raise MemoryError)
-        or 0 otherwise.
-
-        Parameters
-        ----------
-        start : SIZE_t
-            The index of the first sample to consider
-        end : SIZE_t
-            The index of the last sample to consider
-        weighted_n_node_samples : ndarray, dtype=double pointer
-            The total weight of those samples
-        """
-
-        self.start = start
-        self.end = end
 
         self.criterion.init(
             self.y,
             self.sample_weight,
             self.weighted_n_samples,
             self.samples,
-            start,
-            end
+            self.start,
+            self.end
```

Comment:
Aren't the values of `self.start` and `self.end` not assigned yet?

Reply:
This was the big question I had for you @adam2392! Originally when this was in the `node_reset`, it was just using class variables (like `start` and `end`) as they were passed as arguments. When I brought this function out after referring to your note, I realized that there was no reference of `start` and `end`, so I changed them. Later, I actually thought that this `criterion.init` wasn't needed as the original `Splitter` didn't have it. So I experimentally took this out, but when I did, it fails to even build. I'll change it back to plain class variables and see how it changes, but it'd be awesome if you could shed some light on this

Reply:
Oh, if you look upstream, then Tree calls `splitter.init` and `splitter.node_reset`. I would just remove `start` and `end` from `criterion.init` because init should just be initializing the data. 

Tree has to call `splitter.node_reset` anyways at the start of the tree building process to process the `start` and `end` pointers. This will then call `criterion.pointer_reset`. 

So the end code should look like:
```
Splitter:
    cdef init(...):
         # initialize data
         criterion.init(self.y,
             self.sample_weight,
             self.weighted_n_samples,
             self.samples)

   cdef node_reset(...):
         criterion.pointer_reset(start, end)
```

Please update the pxd and pyx files accordingly.

The only issue is that this changes the function signature of `Criterion.init`. I think this change makes sense semantically, but it would be nice to have @jjerphan weigh in on this proposed change.

Reply:
Thank you @adam2392! This helped me better understand the whole code structure 😃. I've pushed new commits reflecting this proposal for @jjerphan's review